api/envoy/config/core/v3/base.proto

```diff
@@ -329,6 +329,23 @@ message HeaderValue {
   ];
 }
 
+// Header name/value pair with header value as bytes.
+message HeaderValueBytes {
+  option (udpa.annotations.versioning).previous_message_type = "envoy.api.v2.core.HeaderValue";
+
+  // Header name.
+  string key = 1
+      [(validate.rules).string =
+           {min_len: 1 max_bytes: 16384 well_known_regex: HTTP_HEADER_NAME strict: false}];
+
+  // Header value.
+  //
+  // The same :ref:`format specifier <config_access_log_format>` as used for
+  // :ref:`HTTP access logging <config_access_log>` applies here, however
+  // unknown header values are replaced with the empty string instead of ``-``.
+  bytes value = 2 [(validate.rules).bytes = {min_len: 0 max_len: 16384}];
```

Comment:
Did you think of adding a `value_bytes` field to the existing `HeaderValue` and treating it as a one-off. The advantage is that we could start to upgrade existing APIs in-place. For ext_proc situation, I think it might be OK to make this change given the WiP statement, although @mattklein123 has some concerns that these APIs are now in heavy use.

Reply:
Upgrade of APIs in place can be problematic. How would a client know if a server supports reading from both fields?

Reply:
If the entire API is still marked alpha/WiP, I guess it's OK, but I do get the impression that people are using this thing. I would have this be the last change and then mark it non-alpha, or just mark it non-alpha now and start with the standard API change semantics that we do for other stable APIs.

Reply:
Thanks for the comments!

I just added the corresponding change in the ProcessingResponse proto message to make non-utf8 character supported in both directions (Envoy <---> ext_proc server) .

Regarding should we add a separate message,  or just add a one-of field, I think one-of is a good idea.  It avoids defining largely duplicated messages, like HeaderValueOption ---> HeaderValueOptionBytes.  But I don't fully understand why it has the advantage of upgrading APIs in place. Could you please elaborate @htuch ?  IIUC, either approach is a breaking change, which needs ext_proc filter and server upgraded at same time with a little bit downtime.

cc @adisuissa for his opinion.

Reply:
You could populate both fields for a period of time, under runtime control, to avoid ext_proc / server needing to upgrade at same time.

Reply:
Right. Yeah, it would be useful to get @yanavlasov and @envoyproxy/api-shepherds take on this, it's a pretty fundamental change if we decide to extend `HeaderValue` this way (with `value_bytes`). The tradeoffs I see is it allows us to retrofit proper header value encoding for UTF-8. The downside is that it's a shared proto used in many places and it wil lbe hard to tell which ones are using the bytes vs. string representation.

Reply:
> You could populate both fields for a period of time, under runtime control, to avoid ext_proc / server needing to upgrade at same time.

This is a de-facto a deprecation and removal of the old field. The alternative of populating both fields is worse IMHO (large message-sizes).

> I guess adding runtime flag applies to other filters, like ext_authz, if they plan to move to the value_bytes in case they want to fix the non-utf8 character issue. For the ext_proc filter code here, we can directly populate the value_bytes since it is in Alpha state, right?

I'm not sure it is possible, for example if the ext_proc server still uses the old proto definition, i won't be able access these fields.

An alternative could be to add a some encoding/decoding ability to send the value as a string. This can be accompanied with some field that signals when the sent string is encoded.

Reply:
Thanks @htuch . The downside you mentioned does raise a concern to me. In that regards, using a separate message can make the change impact localize to the ext_proc  itself, which probably is better solution for the short-term goal?.

Reply:
@adisuissa  the alternative you mentioned by encoding a non-utf8 character into string before sending it is still a breaking change with some complexity.  If we can make breaking change, send it as bytes seem to be simpler.

Reply:
I changed the code to have ext_proc by default support non-utf8 character encoding since IMHO this is a bug fix. For customers who want to maintain the old behavior, they can toggle the run time flag to be false.